sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/util/MergingActiveWindowSet.java

```diff
@@ -175,13 +190,28 @@ public boolean isActive(W window) {
   }
 
   @Override
-  public void addNew(W window) {
+  public void seenWindow(W window) {
     if (!windowToActiveWindow.containsKey(window)) {
+      Preconditions.checkState(!activeWindowToStateAddressWindows.containsKey(window));
       activeWindowToStateAddressWindows.put(window, new LinkedHashSet<W>());
+      windowToActiveWindow.put(window, window);
     }
   }
 
   @Override
+  public void usingActiveWindow(W window) {
+    Set<W> stateAddressWindows = activeWindowToStateAddressWindows.get(window);
+    Preconditions.checkState(stateAddressWindows != null, "Window %s is not ACTIVE or NEW", window);
+    if (stateAddressWindows.isEmpty()) {
+      // This window was NEW but since it survived merging must now become ACTIVE.
```

Comment:
I think this comment belongs in ReduceFnRunner, since it is not about the behavior of this method (making a NEW window ACTIVE) but about the fact that the window survived merging, and therefore this is the right behavior.